Implementation of Checks on Charging and Discharging Voltage Curves in Kinetic Battery Model

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8817 
 - It was noticed that when using the kinetic battery model that if the charging energy and the discharging energy over an annual simulation for a battery that started with nothing in storage did not make sense.  It was reporting that the discharging energy was higher than the charging energy, something that should not occur.  The model equations were checked thoroughly, and the code was correctly calculating the storage in amp-hours.  However, since the voltage varies in this model, the problem was that due to those differences, the battery was able to take out more energy than was put in.  So while amp-hours were being conserved correctly, energy in Joules was not being conserved.  Now, a check on the charging and discharging voltage curves is made during the get input routine.  The new check looks at various fraction of full storage and compares the resulting voltage from the charging and discharging curves.  If discharging is higher in voltage, this is flagged with a warning message since a higher discharging voltage means that the energy coming out of the battery is higher than what went in.  These checks are only done once at the end of the get input so as to not slow down the simulation. A unit test focused on the new subroutine checking the curves is included as well as documentation mods.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Changes for New Battery Variables

Draft code for new variables in the kinetic battery model as well as logic.  This is a draft version until the issue with the IDD is resolved.  Once that is figured out, the rest of the new code (currently commented out) will be instituted.

RKStrand committed:
Fix Format Issue for Kinetic Battery

Format issue that needed to be addressed.

RKStrand committed:
Unit Test and Docs

With the IDD now being processed correctly, code that had been commented out temporarily was reinstated for full functionality.  This also includes a unit test and modified documentation for the new input and output fields.

RKStrand committed:
Kinetic Battery Format Check Issue

Not sure how this passed before and not now, but this addresses the issue.

RKStrand committed:
More Battery Check Stuff

This was an mistake but I'm not sure why it wasn't triggering this before...

rraustad commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3512,6 +3512,12 @@ ElectricStorage::ElectricStorage( // main constructor
+                conserveBattery_ = BatteryEnergyConservationType::CurrentAndEnergy;
+            } else {
+                conserveBattery_ = BatteryEnergyConservationType::Basic;
+            }
```
If you have an enum you should just look it up with getEnumerationValue (soon to be getEnumValue) using a new constexpr string_view array. But that may not be necessary since I don't like these options. Basic does not conserve energy nor does CurrentAndEnergy. So you will have to explain why conservation of energy is not possible with this model. I have not looked at the charge and discharge curves but I would expect a SOC and voltage relationship. Why can't a charge or discharge cycle know exactly how much energy was transferred? @EnergyArchmage may know more about this model?

RKStrand replied:
I will try my best to explain what I am seeing in this model as a potential issue.  During charging, the model calculates a new level of "state of charge" in amp-hours based on the current.  It's got two pots--one for available and one for bound.  The sum of those two equals the amount stored in the battery.  Based on the current and voltage, it also calculates how much energy in Joule is charged or discharged using basically a form of current times voltage.  The voltage is a function of the fraction charged.  During discharging, the same is basically true except that there are two correlations for voltage--one for charging and one for discharging.  If the voltage calculated during discharging is higher than the voltage during charging at the same state of charge and current, then it is possible that since energy is based on current times voltage that the discharging energy is greater than the charging energy to get to that point, purely because the correlation between fraction charged and voltage result in different values.  I am tracking both amp-hours and current in this attempt at a fix and zeroing out amp-hours when energy gets to zero.  This avoids taking out more than one puts in.  There are potentially other approaches--like looking at the charging and discharging curves, comparing the two and notifying the user if they might run into problems and what the problem would look like.  I realize that I failed to add additional documentation to explain this when I pushed this and turned it into a pull request.  I can definitely correct that if we stick with what I proposed.  However, if there are still concerns about that and we want to go with an approach that looks at the curve up front or something else, I am open to discussing that.

EnergyArchmage replied:
It does sound like bad input curves are the problem here.  Checking the curves up front seems like a helpful thing to do.  

I wonder if the output variable Electric Storage Total Voltage could be examined during times discharging and compared to times when charging to identify/explain the problem.

RKStrand replied:
Ok, it seems like it a bad input curve problem.  I looked at the voltage when in charging mode.  The max value is 33.8 V and the average (when not zero) is 33.3 V.  For discharging, the max value is 39 V and the average is 38.7 V.  Since discharging voltage is always higher than charging, that's where it is creating energy.  So, perhaps a better approach would be to check the curves after and notify the users when discharging is higher than charging.  I can create a subroutine that goes through a range of values and compares the two.  If discharging is higher than charging, then a warning would be produced.  I can leave the new output variables in there, but get rid of the new input and the code that processes it and then uses it.  Before I move forward with that, I'd like to hear back that yes that seems reasonable from both @rraustad and @EnergyArchmage.  That way, I'm getting it right (the second time).  Thanks!

EnergyArchmage replied:
That seems reasonable to me.

rraustad replied:
Yes that seems reasonable. And you can get rid of the new input field for Battery Conservation Method. I'm not sure if it's useful to have a starting SOC input as Initial Energy Stored in Battery but keeping or deleting that input field is up to you. What was the original assumption for SOC?

RKStrand replied:
There is an input for SOC in amp-hours.  I think I will just get rid of the extra bookkeeping for storage in Joules.  Essentially, the program will now alert the user when discharging voltage is higher than charging voltage and what to do to resolve that.  I'll add some text to the IO Ref to alert the users about that also.  I'll let you both know when I've implemented the new approach and it's ready for another look.  Thanks to both of you for your help on this!

RKStrand replied:
I should have also said that I'll get rid of both of the new input fields.

rraustad replied:
Is this a check of the curves in getInput or a recurring check during the simulation? I'd rather see a check of the curves in getInput.

RKStrand replied:
Definitely during the getInput routine.  It needs to check different charge fractions so that discharging voltage stays below charging voltage for any charged fraction.  If I do that upfront, then there should not be any need to do it during the actual simulation where it could just end up being expensive overhead over an annual simulation.

RKStrand replied:
Ok, this is ready to be looked at again.  Tests are still churning, but I don't anticipate any red flags.  I got rid of what was done before and replaced it with a subroutine that gets called during the get routine.  Goes through a range of values for the charging and discharging curves.  If discharging voltage is higher than charging, it triggers a warning message.  The test file that is already in the test suite, ShopwithPVBattery.idf, actually triggers the warning now, which it should given the values of the parameters.  It's a short run so it doesn't end up showing the discharging energy is higher than charging.  However, if you switch to a full annual run, discharging is higher.  I haven't changed the idf yet.  If we feel that it would be better to have a file that doesn't trip this warning message, I'm ok with that.  Happy reviewing!

RKStrand committed:
Back out changes made to code

After discussion during review of what was done, it was decided to go in a different direction.  This commit undoes almost all of the changes other than a few minor, random documentation fixes.  This will provide a new base from which to start for the new solution to this defect.

RKStrand committed:
Revised Solution for Kinetic Battery Issue

After undoing the previous solution, a new solution which checks the charging and discharging curves before the simulation is implemented with this push.  If charging voltage is less than discharging, a warning gets thrown.

RKStrand committed:
Merge branch 'develop' into 8817KineticBatteryOutputVariableAddition

RKStrand committed:
Kinetic Battery: correction to format issue

The custom format check didn't like something that Xcode was happy to just deal with.  Also modified some of the warning messages.

RKStrand committed:
Revised Unit Test and Docs Mods

New unit test and documentation edits for the revised solution.

rraustad commented:
I'm going to suggest a different err message.

    +   ** Warning ** Kinetic Battery Model: KIBAM has a charging/discharging voltage curve conflict.
    +   **   ~~~   ** Discharging voltage is higher than charging voltage which may potentially lead to an imbalance in the stored energy.
    +   **   ~~~   ** Check the charging and discharging curves to make sure that the charging voltage is greater than discharging.
    +   **   ~~~   ** Also check the charging and discharging energy outputs to find any discrepancies.
    +   **   ~~~   ** Charged fraction = 0.0, Charging voltage = 12.400 V, Discharging voltage = 12.591 V
    +   **   ~~~   ** Charged fraction = 0.1, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.2, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.3, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.4, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.5, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.6, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.7, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.8, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 0.9, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V
    +   **   ~~~   ** Charged fraction = 1.0, Charging voltage = 12.xxx V, Discharging voltage = 12.xxx V

rraustad commented on tst/EnergyPlus/unit/ElectricPowerServiceManager.unit.cc:
```diff
@@ -1230,3 +1230,148 @@ TEST_F(EnergyPlusFixture, Battery_checkUserEfficiencyInputTest)
+
+TEST_F(EnergyPlusFixture, Battery_checkChargeDischargeVoltageCurves)
+{
+    std::string battName;
```
battName doesn't look like it's used.

RKStrand replied:
I'll get rid of it.  Forgot that it was there when I just assigned the name in-line.

Also, just for clarity (not sure why it's not letting me comment on this above), the reworded error message--sounds good to me, but are you advocating for only doing the check every tenth or only reporting on the tenths when there is a problem?

rraustad replied:
I think the high resolution checks are good but if there is a problem report out the low resolution results. The original message was too bulky. You may have other ideas of how to briefly report out the issue.

RKStrand replied:
Ok, so what I do now is I do the higher resolution for checks but as soon as it finds a problem it jumps out and then reports on the tenths.  Unless I accidentally missed something, I used your suggested wording.  It looked good to me and was much less bulky.  Commit coming up soon...

RKStrand replied:
Commit made, merged in develop, and pushed this all up to GitHub.  Let me know if there is anything else that needs to be addressed.  Thanks!

rraustad commented on tst/EnergyPlus/unit/ElectricPowerServiceManager.unit.cc:
```diff
@@ -1230,3 +1230,148 @@ TEST_F(EnergyPlusFixture, Battery_checkUserEfficiencyInputTest)
+    checkChargeDischargeVoltageCurves(*state, "Test 2", e0c, e0d, chaCurveIndex, disCurveIndex);
+
+    std::string const error_string2 = delimited_string(
+        {"   ** Warning ** Kinetic Battery Model: Test 2 has a charging/discharging voltage curve conflict.",
```
You should probably add the battery name in case there are multiple batteries.

rraustad replied:
Oh, `Test 2` is the battery name.

RKStrand replied:
Yup--Test 2 is the battery name.  The only reason it is passed is for better reporting if there is a problem.

RKStrand replied:
I've changed the name to "Battery for Test 2" so it's clearer hopefully.

rraustad commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1326,8 +1334,8 @@ \subsubsection{Outputs}\label{electricloadcenter-storagebattery-outputs}
-  HVAC,Average,Electric Storage Charge State{[}Ah{]}
-\item
+  HVAC,Average,Electric Storage Charge State{[}Ah{]}  
+\\item
```
What happened here?

RKStrand replied:
Oops...I forgot to rip that out when I backed things up.  There were a couple of documentation minor edits that I didn't want to lose and then forgot about removing these from the list.  I'll get rid of that/those three.

RKStrand replied:
Oops, part 2...I got rid of the old ones but accidentally missed a "\".  Sorry--made a bad assumption without looking carefully at what you referenced.  Anyway, it's gone in the next commit.  Good catch!

rraustad commented on src/EnergyPlus/ElectricPowerServiceManager.hh:
```diff
@@ -833,6 +833,9 @@ void createFacilityElectricPowerServiceObject(const EnergyPlusData &state);
 
+void checkChargeDischargeVoltageCurves(
+    EnergyPlusData &state, std::string const nameBatt, Real64 const E0c, Real64 const E0d, int const chargeIndex, int const dischargeIndex);
+
```
std::string_view is desired. std::string arguments are referenced, std::string_view are not.

RKStrand replied:
I'll change that in the next commit.

rraustad replied:
Arguments look good now.

Myoldmopar replied:
:+1:

rraustad reviewed (changes requested):
Only a few comments, otherwise this looks good.

RKStrand committed:
Kinetic Battery Address Review Comments

Addressed various review comments received for this bug fix.

RKStrand committed:
Merge branch 'develop' into 8817KineticBatteryOutputVariableAddition

rraustad commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1326,7 +1334,7 @@ \subsubsection{Outputs}\label{electricloadcenter-storagebattery-outputs}
   HVAC,Average, Electric Storage Operating Mode Index {[]}
 \item
-  HVAC,Average,Electric Storage Charge State{[}Ah{]}
+  HVAC,Average,Electric Storage Charge State{[}Ah{]}  
```
It looks like there are 2 spaces at the end of this line. CI Clang will catch this if true and can be corrected on merge. No need for another commit to fix this if it is the only change.

rraustad commented on tst/EnergyPlus/unit/ElectricPowerServiceManager.unit.cc:
```diff
@@ -1230,3 +1230,103 @@ TEST_F(EnergyPlusFixture, Battery_checkUserEfficiencyInputTest)
+        "   **   ~~~   ** Charged fraction = 1.0, Charging voltage = 12.410 V, Discharging voltage = 12.600 V",
+    });
+    EXPECT_TRUE(compare_err_stream(error_string2, true));
+}
```
And more compact and descriptive message.

Myoldmopar replied:
:+1:

rraustad reviewed (approved):
This looks ready for other review or merge.

EnergyArchmage commented:
Not critical, but I think the example file should be corrected with curves that work okay.  I no longer have access to the references etc.  I believe the kinetic battery model is also in SAM, maybe someone from NREL residential group could find different curve inputs from SAM libraries?

RKStrand commented:
Looks like there is an unused variable that I need to get rid of...so I'll make another commit and take care the extra spaces also.  For the input file, I can use the coefficients that are used in the test case (Battery for Test 1).  Those work though there is a lot more variation in the voltages.  I could always massage the ones that are in there and get a better result.  I'll try the small adjustments first and see what will work and still be relatively close to what is in there now.  I'll work on this in a little bit and then make another commit.

RKStrand committed:
Final Tweaks to Kinetic Battery Fix?

A few more minor edits to various files to clean up some other issues noted in review comments and in the last commit runs.

RKStrand committed:
Merge branch 'develop' into 8817KineticBatteryOutputVariableAddition

RKStrand commented:
Ok, I just committed another round of small edits and merge in develop again.  Hopefully this caught everything.  If not, let me know and I'll fix it.  Thanks again for the reviews, @rraustad and @EnergyArchmage!

nrel-bot-2b commented:
@RKStrand it has been 26 days since this pull request was last updated.

nrel-bot-2 commented:
@RKStrand it has been 7 days since this pull request was last updated.

Myoldmopar commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1154,7 +1154,7 @@ \subsection{ElectricLoadCenter:Storage:Battery}\label{electricloadcenterstorage-
 Each individual battery module is modeled as a voltage source in series with an electrical resistance. KiBaM assumes that the internal resistance is constant and the open circuit voltage varies with the electric current, the state of charge and the operation mode (charging or discharging). For an individual battery module, the open circuit voltage at any time is correlated to the voltage at fully ...
 
-The object offers user the option to perform battery life calculation. If battery life is modeled, the user needs to provide a group of coefficients for the correlation between the number of cycles for battery failure and the corresponding cycle range. More detailed information can be found from the Engineering Reference.
+The object offers users the option to perform battery life calculation. If battery life is modeled, the user needs to provide a group of coefficients for the correlation between the number of cycles for battery failure and the corresponding cycle range. More detailed information can be found from the Engineering Reference.
```
Yep, fine.

Myoldmopar commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1184,7 +1184,7 @@ \subsubsection{Inputs}
 \paragraph{Field: Maximum Module Capacity}\label{field-maximum-module-capacity}
 
-This field indicates the maximum capacity of one battery module. It is evaluated as the total Amp-hours available when a full battery is discharged at infinitesimal current. This capacity is for an individual battery module, not for the whole battery bank. The maximum capacity can be found from manufacture s data or derived from test data by curve fitting.
+This field indicates the maximum capacity of one battery module. It is evaluated as the total Amp-hours available when a full battery is discharged at infinitesimal current. This capacity is for an individual battery module, not for the whole battery bank. The maximum capacity can be found from manufacture's data or derived from test data by curve fitting.
```
I'm not sure `manufacture's` is better than `manufacture s`, but I"m not interested in holding up the PR for more commits.

Myoldmopar commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1216,7 +1216,11 @@ \subsubsection{Inputs}
+
+Note that a check is made when the input is retrieved to look at the voltage when charging and discharging (see next field). The voltage during charging at a particular battery charge should be higher for charging.  If not, it would potentially lead to results that show more energy being discharged by the battery than was charged. Warning messages are produced to alert the user of this situation  ...
+
+More details on the kinetic battery as well as the charging and discharging algorithm can be found from the Engineering Reference.
```
:+1:

Myoldmopar commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1228,7 +1232,11 @@ \subsubsection{Inputs}
+
+Note that a check is made when the input is retrieved to look at the voltage when charging (see previous field) and discharging. The voltage during charging at a particular battery charge should be higher for charging.  If not, it would potentially lead to results that show more energy being discharged by the battery than was charged. Warning messages are produced to alert the user of this situat ...
+
+More details on the kinetic battery as well as the charging and discharging algorithm can be found from the Engineering Reference.
```
:+1:

Myoldmopar commented on doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:
```diff
@@ -1267,57 +1275,57 @@ \subsubsection{Inputs}
-  Yes,                                   !- Battery Life Calculation
-  10,                                   !- Number of Cycle Bins
-  BatteryCycleCurve;       !- Battery Cycle Life Curve Name
+  Lead Acid Battery A,   !- Name
```
A fine formatting cleanup :broom:

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -92182,7 +92182,7 @@ ElectricLoadCenter:Storage:Battery,
         \default 10
         \note Only required when battery life calculation is activated
-   A7; \field Battery Life Curve Name
+   A7;  \field Battery Life Curve Name
```
OK, so while it technically is an IDDChange, it's not a functional IDD change.  Got it.

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3839,6 +3843,45 @@ Real64 checkUserEfficiencyInput(EnergyPlusData &state, Real64 userInputValue, st
+void checkChargeDischargeVoltageCurves(
+    EnergyPlusData &state, std::string_view nameBatt, Real64 const E0c, Real64 const E0d, int const chargeIndex, int const dischargeIndex)
+{
+    int constexpr numChecks = 50;  // number of divisions for checking the functions from 0 to 1 for fraction charged/discharged
```
Constexpr woot!

Myoldmopar commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3839,6 +3843,45 @@ Real64 checkUserEfficiencyInput(EnergyPlusData &state, Real64 userInputValue, st
+        Real64 xfc = float(loop - 1) / float(numChecks);                               // = q0/qmax
+        Real64 xfd = 1.0 - xfc;                                                        // = (qmax-q0)/qmax = 1 - xfc
+        Real64 chargeVoltage = E0d + Curve::CurveValue(state, chargeIndex, xfc);       // E0d+Ac*xfc+Cc*xfc/(Dc-xfc)
+        Real64 dischargeVoltage = E0c + Curve::CurveValue(state, dischargeIndex, xfd); // E0c+Ad*xfd+Cd*xfd/(Dd-xfd)
```
I would have given a bonus point for making these `const`

Myoldmopar commented on testfiles/ShopWithPVandBattery.idf:
```diff
@@ -5080,7 +5080,7 @@
     0.0899,                  !- Coefficient1 C1
     -98.24,                  !- Coefficient2 C2
-    -.0082,                  !- Coefficient3 C3
+    -.82,                    !- Coefficient3 C3
```
I would assume this will cause diffs

Myoldmopar replied:
And it looks like it does.  No problem.

Myoldmopar commented:
Just rebuilt and tested this again and everything looks good.  One diff in the IDF that has input changes.  Merging this one in, thanks @RKStrand
